routes/adjustmentMethods/EToAdjustmentMethod.ts

```diff
@@ -0,0 +1,224 @@
+import * as SunCalc from "suncalc";
+import * as moment from "moment";
+import { AdjustmentMethod, AdjustmentMethodResponse, AdjustmentOptions } from "./AdjustmentMethod";
+import { GeoCoordinates, WateringData, WeatherProviderId } from "../../types";
+import { WeatherProvider } from "../weatherProviders/WeatherProvider";
+
+
+/**
+ * Calculates how much watering should be scaled based on weather and adjustment options by comparing the recent
+ * potential ETo to the baseline potential ETo that the watering program was designed for.
+ */
+async function calculateEToWateringScale(
+	adjustmentOptions: EToScalingAdjustmentOptions,
+	wateringData: WateringData | undefined,
+	coordinates: GeoCoordinates,
+	weatherProvider: WeatherProvider
+): Promise< AdjustmentMethodResponse > {
+
+	// Temporarily disabled since OWM forecast data is checking if rain is forecasted for 3 hours in the future.
+	/*
+	if ( wateringData && wateringData.raining ) {
+		return {
+			scale: 0,
+			rawData: { raining: 1 }
+		}
+	}
+	 */
+
```

Comment:
I think it would make more sense to validate the required fields in the `WeatherProvider` to ensure that the values they return are as documented), but doing so would mean that `RainDelayAdjustmentMethod` and `ManualAdjustmentMethod` would error if humidity or temperature were missing (even though these methods don't require these properties). This might change depending on how https://github.com/OpenSprinkler/OpenSprinkler-Weather/pull/62/files/69e8d64f41f9ecd3a6e1a6d6ba88fa9791b627ad#r294479850 is resolved, so let's wait until that's been addressed to figure out how to best address this issue.

Reply:
Very good summary. I agree, this and the linked comment are the crux of the issue.

I would prefer for `EToAjustmentMethod` to error if the PWS `WeatherProvider` didn't support a radiation sensor rather than `RainDelayAdjustmentMethod` failing when all the required data is available. Or better yet `EToAdjustmentMethod` should fall back to an approximation calculation if possible i.e. derive radiation from Tmin/Tmax.

At the end of the day, a Weather Provider is a "dumb" device whereas the Adjustment Method is "smart" logic. In general, responsibility for determining how to handle edge cases gravitates to the "smart" code. I would want to try and avoid having to update all the `WeatherProvider`s just because we add a new `AdjustmentMethod` with some esoteric data requirement.

Reply:
> I would prefer for EToAjustmentMethod to error if the PWS WeatherProvider didn't support a radiation sensor rather than RainDelayAdjustmentMethod failing when all the required data is available.

This shouldn't be a concern. Since `EToAdjustmentMethod` and `RainDelayAdjustmentMethod` use different data types (`EToData` vs. `WateringData`), an error would only be thrown if one of the fields in `WateringData` was missing. Each data type defines the minimum information necessary (or at least it should - the Manual and RainDelay methods only need the `precip` field in `WateringData`, but the proposed fix to the linked comment would resolve this issue by creating a `BaseWateringData` type that wouldn't contain any unnecessary fields.

> I would want to try and avoid having to update all the WeatherProviders just because we add a new AdjustmentMethod with some esoteric data requirement.

For similar reasons to the point above, this shouldn't be an issue. The way the code is currently setup, each AdjustmentMethod is independent of each other and adding a new method shouldn't interfere with existing methods as long as we use a new data type for the esoteric method instead of trying to repurpose an existing data type that isn't an appropriate match.

Reply:
Resolving in favor of Slack conversation